### PR TITLE
Specification and Interpreter for  transpose op

### PR DESCRIPTION
**Pull request by sdasgup3**
fixes https://github.com/openxla/stablehlo/issues/171

Diff based on https://github.com/openxla/stablehlo/pull/167

### CODE REVIEW

File: docs/status.md [hunk 1]
```diff
@@ -135,7 +135,7 @@ one of the following tracking labels.
 | TanhOp |yes||||no|
 | TorchIndexSelectOp |no||||no|
 | TraceOp |no||||no|
-| TransposeOp |no||||no|
+| TransposeOp |yes||||yes|
```
Comment (burmako) [inline docs/status.md:138]: Let's also check whether TransposeOp's verifier and shape function conform to the newly written spec. If they do, let's flip `yes*` in the table to `yes`. If they don't let's open tickets about this.

File: stablehlo/tests/interpret_transpose.mlir [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+  // CHECK-NEXT:   28 : i32
+  // CHECK-NEXT:   29 : i32
+  // CHECK-NEXT:   30 : i32
+  // CHECK-NEXT:   31 : i32
+  // CHECK-NEXT:   32 : i32
+  // CHECK-NEXT:   33 : i32
+  // CHECK-NEXT:   34 : i32
+  // CHECK-NEXT:   35 : i32
+  // CHECK-NEXT:   36 : i32
+}
```
Comment (burmako) [inline stablehlo/tests/interpret_transpose.mlir:114]: This test doesn't check anything new and looks fairly laborious for a human to follow. Perhaps delete?

File: docs/spec_draft.md [hunk 3]
```diff
@@ -1029,6 +1030,52 @@ specification. Numeric precision is implementation-defined.
...
 
+## stablehlo.transpose
+
+`stablehlo.transpose(operand, permutation) -> result`
+
+### Semantics
+
+Permutes the dimensions of `operand` tensor using a permutation of its shape
+`permutation` and produces a `result` tensor. The $i^{th}$ dimension of the
+`result` is given by `permutation`$[i]^{th}$ dimension of the `operand`.
```
Comment (burmako) [inline docs/spec_draft.md:1041]: This section doesn't really explain how to obtain the `result` tensor from the `operand` tensor. "The X dimension of a tensor is given by the Y dimension of another tensor" sounds fancy but is not actionable.

Here, I have the same feedback as in #170. Let's define op semantics from first principles, using only the definitions in this spec and college maths. Before proceeding with that, let's first expand our vocabulary - something which I took a stab at in #193.

Comment (burmako) [inline docs/spec_draft.md:1041]: With the terminology from #193, this section could be rewritten as follows:

```
Permutes the dimensions of `operand` tensor using `permutation` and produces a 
`result` tensor. More formally, `result[i0, ..., iR-1]` is equal to 
`operand[j0, ..., jR-1]` where `i[d] = j[permutation[d]]`.
```

File: docs/spec_draft.md [hunk 4]
```diff
@@ -1029,6 +1030,52 @@ specification. Numeric precision is implementation-defined.
...
+
+### Operands
+
+| Name | Type |
+|-|-|
+| `operand` | tensor of any supported types |
+| `permutation` | array of type `ui64` |
```
Comment (burmako) [inline docs/spec_draft.md:1048]: For better or for worse, this is currently an array of `si64`, so let's start with reflecting the reality, and in the future we'll be able to change this to `ui64` if we decide.

File: docs/spec_draft.md [hunk 5]
```diff
@@ -1029,6 +1030,52 @@ specification. Numeric precision is implementation-defined.
...
+### Results
+
+| Name | Type |
+|-|-|
+| `result` | tensor of any supported types |
+
+### Constraints
+
+  * (C1) `operand` and `result` have the same element type.
+  * (C2) `permutation` is a permutation of `operand`'s shape.
```
Comment (burmako) [inline docs/spec_draft.md:1059]: "* (C2) \`permutation\` is a permutation of \`[0, 1, ..., R-1]\` where \`R\` is the rank of \`operand\`".

Comment (burmako) [inline docs/spec_draft.md:1059]: "* (C3) \`result\`'s shape is a permutation of \`operand\`'s shape".

File: stablehlo/reference/Ops.cpp [hunk 6]
```diff
@@ -47,5 +47,37 @@ Tensor eval(SineOp op, const Tensor &operand) {
...
+
+  // The operation itself does not require any copying but involves swapping
+  // strides.
+  std::vector<int64_t> stride(operand.getStrides());
+  std::vector<int64_t> permutation(perm.getValues<int64_t>().begin(),
+                                   perm.getValues<int64_t>().end());
+  applyInPlacePermutation(stride, permutation);
+  result.setStrides(stride);
+  return result;
+}
```
Comment (burmako) [inline stablehlo/reference/Ops.cpp:80]: Similarly to #170,  let's define op implementations from first principles, i.e. using only the definitions in the spec and APIs from `Tensor.h` and `Element.h` which implement these definitions. 

From that perspective, relying on exact representation of a tensor, i.e. something that the spec treats as implementation-defined (currently, this is implicit, #193 makes this explicit), doesn't sound right.

I think improving on this is fairly straightforward if we use the iterator that you proposed in https://github.com/openxla/stablehlo/pull/167 (with the modifications that I proposed in the comments), for example:

```
Tensor eval(TransposeOp op, const Tensor &operand) {
  Tensor result(op.getType());
  for (auto resultIt = result.index_begin(); resultIt != result.index_end(); ++resultIt) {
    auto resultIndex = *resultIt;
    auto operandIndex = permute(resultIt, op.permutation());
    result.set(resultIndex, operandIndex);
  }
  return result;
}
```

File: stablehlo/reference/Ops.cpp [hunk 7]
```diff
@@ -47,5 +47,37 @@ Tensor eval(SineOp op, const Tensor &operand) {
   return result;
 }
 
+namespace {
+
+// Appies permutation `perm` to an array `array` where perm[i] indicates the
+// location where the current array[i] goes.
+void applyInPlacePermutation(std::vector<int64_t> &array,
+                             const std::vector<int64_t> &perm) {
```
Comment (burmako) [inline stablehlo/reference/Ops.cpp:55]: Is there a strong reason to apply the permutation in place? Functional style would be much easier to understand, e.g.:

```
llvm::SmallVector<int64_t> permute(ArrayRef<int64_t> array, DenseIntElementsAttr permutation) {
  llvm::SmallVector<int64_t> result(array.size());
  for (auto i = 0; i < result.size(); ++i) {
    result[permutation.getValues<int64_t>()[i]] = array[i];
  }
  return result;
}
```

File: stablehlo/reference/Ops.cpp [hunk 8]
```diff
@@ -47,5 +47,37 @@ Tensor eval(SineOp op, const Tensor &operand) {
...
+  for (size_t i = 0; i < perm.size(); i++) {
+    swapIdx = perm[i];
+    while (swapIdx < i) {
+      swapIdx = perm[swapIdx];
+    }
+    std::swap(array[i], array[swapIdx]);
+  }
+}
+
+}  // namespace
```
Comment (burmako) [inline stablehlo/reference/Ops.cpp:66]: Move to the top of the file?

File: docs/spec_draft.md (hunk 4 above)
Comment (burmako) [inline docs/spec_draft.md:1048]: "1-dimensional array".

**Review by burmako — CHANGES_REQUESTED** [review]

File: docs/spec_draft.md [hunk 9]
```diff
@@ -1029,6 +1030,52 @@ specification. Numeric precision is implementation-defined.
...
+
+### Examples
+
+```mlir
+// %operand: [
+//            [[1,2], [3,4], [5,6]],
```
Comment (ghpvnist) [inline docs/spec_draft.md:1065]: Let's add spaces after comma to be consistent with the rest of the doc here and below.

File: docs/status.md [hunk 10]
```diff
@@ -75,7 +75,7 @@ one of the following tracking labels.
 | DynamicUpdateSliceOp |no||||no|
 | EinsumOp |no||||no|
 | Expm1Op |no||||no|
-| ExpOp |no||||no|
+| ExpOp |yes||||no|
```
[the remark this answers is not in the document]
Comment (sdasgup3) [reply docs/status.md:78]: Sure.

File: stablehlo/tests/interpret_transpose.mlir (hunk 2 above)
Comment (sdasgup3) [reply stablehlo/tests/interpret_transpose.mlir:114]: deleted.

File: stablehlo/reference/Ops.cpp (hunk 7 above)
Comment (sdasgup3) [reply stablehlo/reference/Ops.cpp:55]: The in-place permute is not needed. Replaced it with a simpler implementation.

File: docs/spec_draft.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (sdasgup3) [reply docs/spec_draft.md:1041]: not applicable with the latest revision.

File: docs/status.md (hunk 1 above)
Comment (sdasgup3) [reply docs/status.md]: done.

File: docs/spec_draft.md [hunk 11]
```diff
@@ -1072,6 +1073,56 @@ Numeric precision is implementation-defined.
...
+
+`stablehlo.transpose(operand, permutation) -> result`
+
+### Semantics
+
+Permutes the dimensions of `operand` tensor using `permutation` and produces a
+`result` tensor. More formally, `result[i0, ..., iR-1]` is equal to
+`operand[j0, ..., jR-1]` where `i[d] = j[permutation[d]]`.
+
+
```
Comment (burmako) [inline docs/spec_draft.md:1086]: Extraneous newline.

File: docs/spec_draft.md [hunk 12]
```diff
@@ -1072,6 +1073,56 @@ Numeric precision is implementation-defined.
...
+| `result` | tensor of any supported types |
+
+### Constraints
+
+  * (C1) `operand` and `result` have the same element type.
+  * (C2) `permutation` is a permutation of `[0, 1, ..., R-1]` where `R` is the
+  rank of `operand`.
+  * (C3) `result`'s shape is a permutation of `operand`'s shape.
+
+
```
Comment (burmako) [inline docs/spec_draft.md:1107]: Same here.

File: docs/status.md [hunk 13]
```diff
@@ -135,7 +135,7 @@ one of the following tracking labels.
 | TanhOp |yes|yes*|yes*|yes|yes|
 | TorchIndexSelectOp |no|no|no|no|no|
 | TraceOp |no|yes*|no|yes|no|
-| TransposeOp |no|yes*|yes*|no|no|
+| TransposeOp |yes|yes|yes*|no|yes|
```
Comment (burmako) [inline docs/status.md:138]: Is it intended for Type Inference to stay at "yes*" here? At a glance, the spec is already implemented correctly: https://github.com/openxla/stablehlo/blob/be42540a288a302b6e4071e7d29c0bf8b911589e/stablehlo/dialect/StablehloOps.cpp#L4983-L5022.

File: stablehlo/reference/Ops.cpp [hunk 14]
```diff
@@ -81,5 +94,17 @@ Tensor eval(TanhOp op, const Tensor &operand) {
   return result;
 }
 
+Tensor eval(TransposeOp op, const Tensor &operand) {
+  Tensor result(op.getType());
+  SmallVector<int64_t> permutation =
+      to_vector(op.permutation().getValues<int64_t>());
```
Comment (burmako) [inline stablehlo/reference/Ops.cpp:100]: Can we avoid `to_vector` and just pass `getValues` directly? (By appropriately changing the signature of `permute`).

**Review by burmako — CHANGES_REQUESTED** [review]

File: docs/status.md (hunk 13 above)
Comment (sdasgup3) [reply docs/status.md:138]: I agree. Updated.

File: stablehlo/reference/Ops.cpp (hunk 14 above)
Comment (sdasgup3) [reply stablehlo/reference/Ops.cpp:100]: Sure. The signature looks verbose, but, being a static function, does not hurt.

**Review by burmako — APPROVED** [review]

**Pull request merged**